### PR TITLE
handle ansi style in cell content

### LINKED ISSUE #91: [Feature] Handle ANSI escape codes from external libraries
#### Describe the bug

Column alignment breaks if printed text has formatting control codes.

#### Expected behavior

Text in a column contains parts formated with color/style using `console` crate. Alignment of later columns should be correct.

#### Code
```
use comfy_table::{Table, Row, Cell};
use console::Style;

fn main() {
    let mut table = Table::new();
    table.load_preset(comfy_table::presets::NOTHING);
    
    let mut row = Row::new();
    row.add_cell(Cell::new(
        "cell1"
    ));
    row.add_cell(Cell::new(
        "cell2"
    ));

    table.add_row(row);

    let mut row = Row::new();
    row.add_cell(Cell::new(
        format!("cell{}", console::style("3").bold().red())
    ));
    row.add_cell(Cell::new(
        "cell4"
    ));
    table.add_row(row);

    println!("{}", table);   
}
```
image

#### Additional context

- Operating System: nixos linux
- comfy-table version: 6.1.0

### PR DESCRIPTION
**Pull request by blueforesticarus**
intends to supersede #77
would close #91

### CODE REVIEW

**Comment by blueForestIcarus** [comment]
@Nukesor  could you approve the CI workflow?

**Comment by Nukesor** [comment]
@blueForestIcarus You can also test the CI on your fork by enabling Actions over there :)

**Comment by blueForestIcarus** [comment]
@Nukesor What do you want the logic to be when the tty feature is not enabled? 
Currently I am using crossterm to identify the reset ansi code, but it looking at the crossterm source it appears to be constant.

**Comment by Nukesor** [comment]
No need to setup codecov :)
As long as the other Actions succeed, you should be good to go.

Regarding the `tty` question, did you introduce any new libraries that contain `unsafe` code? If so, this could be a problem, as people rely on the current dependency tree to be fully safe and non-panicable if `tty` is disabled.

Otherwise, I don't see a big problem with including it in the default feature set.

Did you try the `build_large_tables` benchmark? If the new logic significantly increases the runtime of this benchmark, we should probably think about hiding this logic behind a feature flag (if possible).

**Comment by blueForestIcarus** [comment]
Looks like console has a few unsafe blocks, but its all in term support and nothing I am using touches it.

**Comment by blueForestIcarus** [comment]
```

     Running benches/build_large_table.rs (target/release/deps/build_large_table-81383246e255f825)
Huge table              time:   [18.715 ms 18.734 ms 18.755 ms]
                        change: [+24.626% +25.179% +25.712%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

     Running benches/build_tables.rs (target/release/deps/build_tables-b79d9b1e4fa9749e)
Readme table            time:   [27.808 µs 27.830 µs 27.856 µs]
                        change: [+22.215% +22.616% +23.027%] (p = 0.00 < 0.05)
                        Performance has regressed.

Big table               time:   [454.56 µs 455.41 µs 456.49 µs]
                        change: [+8.5173% +8.9623% +9.3673%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 19 outliers among 100 measurements (19.00%)
  7 (7.00%) high mild
  12 (12.00%) high severe
```

**Comment by blueForestIcarus** [comment]
@Nukesor let me know if that bench result is acceptable or if you want to make a feature flag.

**Comment by Nukesor** [comment]
Sorry for being so slow to respond :sweat: 

I'm currently in the middle of a job transition and really don't have a lot of time to work on my open source projects.
I'll try to get to it as soon as possible :)

File: src/utils/formatting/content_split.rs [hunk 1]
```diff
@@ -28,8 +28,8 @@ pub fn split_line(line: &str, info: &ColumnDisplayInfo, delimiter: char) -> Vec<
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:28]: The current logic allows splits on arbitrary characters, including characters that could be contained inside of ansi escape codes.
We would have to somehow ensure that this won't break

File: src/utils/formatting/content_split.rs [hunk 2]
```diff
@@ -148,45 +148,112 @@ const MIN_FREE_CHARS: usize = 2;
...
-        // The user has to live with this, if they decide to add control characters or some fancy
-        // stuff into their tables. This is considered undefined behavior and we try to handle this
-        // to the best of our capabilities.
-        let character_width = c.width().unwrap_or(1);
+            if !is_esc {
+                //allows popping unneeded escape codes later
+                last_txt = head.len();
+                escape_count = escapes.len();
+            }
+        } else {
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:197]: Early `break` for better readability

File: src/utils/formatting/content_split.rs [hunk 3]
```diff
@@ -148,45 +148,112 @@ const MIN_FREE_CHARS: usize = 2;
...
-        if (current_width + c.width().unwrap_or(1)) > allowed_width {
-            break;
+    let mut iter = console::AnsiCodeIterator::new(word);
+    let mut current_len = 0;
+
+    let mut escapes = Vec::new();
+    let mut head = String::with_capacity(word.len());
+    let mut last_txt = 0;
+    let mut escape_count = 0;
+    let mut tail = String::with_capacity(word.len());
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:173]: Please write proper inline documentation on what this algorithm does.
It's by no means trivial and really hard to read/understand right now.

Try to chunk the algorithm into logical blocks and explain what each block is doing in a sentence or two.

Also, please write docs on what these variables are for.

File: src/utils/formatting/content_split.rs [hunk 4]
```diff
@@ -148,45 +148,112 @@ const MIN_FREE_CHARS: usize = 2;
...
+                current_len += character_width;
+                let c = char_iter.next().unwrap();
+                head.push(c);
+
+                // c is not escape code
+                last_txt = head.len();
+                escape_count = escapes.len();
+            }
+
+            head.truncate(last_txt); // cut off dangling escape codes since they should have no effect
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:215]: Comments above code lines

File: Cargo.toml [hunk 5]
```diff
@@ -42,6 +42,7 @@ crossterm = { version = "0.25", optional = true }
 strum = "0.24"
 strum_macros = "0.24"
 unicode-width = "0.1"
+console = "0.15.1"
```
Comment (Nukesor) [inline Cargo.toml:45]: `console` has a `unicode-width` feature flag.

Without this, multi-character unicode symbols won't be detected as such.

**Comment by Nukesor** [comment]
A ~22%  performance regression for a feature that isn't being used by many users isn't that nice.

I would definitely opt to hide this behind a feature flag!
That way, we can also mark this as experimental and prevent any backward compatibility breakages.

In general, this doesn't look too bad, but it definitely needs a lot of tests :D

**Review by Nukesor — CHANGES_REQUESTED** [review]

File: src/utils/formatting/content_split.rs [hunk 6]
```diff
@@ -148,45 +148,112 @@ const MIN_FREE_CHARS: usize = 2;
...
-    // That is, pushing the next character would result in the string being too long.
-    while let Some(c) = char_iter.peek() {
-        if (current_width + c.width().unwrap_or(1)) > allowed_width {
-            break;
+    let mut iter = console::AnsiCodeIterator::new(word);
+    let mut current_len = 0;
+
+    let mut escapes = Vec::new();
+    let mut head = String::with_capacity(word.len());
+    let mut last_txt = 0;
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:171]: Let's stick with `text`. There's no reason to worsen readability for the sake of a single character

File: src/utils/formatting/content_split.rs [hunk 7]
```diff
@@ -148,45 +148,112 @@ const MIN_FREE_CHARS: usize = 2;
...
+    let mut escape_count = 0;
+    let mut tail = String::with_capacity(word.len());
+
+    for (val, is_esc) in iter.by_ref() {
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:175]: `val` isn't the best name for this variable.

Isn't this a iterator over characters?

File: src/utils/formatting/content_split.rs [hunk 8]
```diff
@@ -148,45 +148,112 @@ const MIN_FREE_CHARS: usize = 2;
...
+        if is_esc {
+            escapes.push(val);
+            if val == ANSI_RESET {
+                escapes.clear();
+            }
         }
 
-        // We can unwrap, as we just checked that a suitable character is next in line.
-        let c = char_iter.next().unwrap();
+        let len = match is_esc {
```
Comment (Nukesor) [inline src/utils/formatting/content_split.rs:183]: What `len`? Be more descriptive

**Comment by blueforesticarus** [comment]
I think I addressed everything

**Comment by Nukesor** [comment]
Nice :)

I'll review it as soon as I find some time to do so!

**Comment by Nukesor** [comment]
First of all, thanks for your work and time!

I'm going to merge this soon, but there were still some open issues I wanted to address first.
Since this already took so long (due to me), I took the liberty to build upon your MR and create a second MR with a few commits on top. I didn't touch any of your commits though!

Feel free to check out my additions over here:
https://github.com/Nukesor/comfy-table/pull/108

**Pull request merged**